tools/chpldef/Message.h

```diff
@@ -0,0 +1,396 @@
+
+  /** Convert a given message tag to string. */
+  static const char* tagToJsonRpcMethodName(Tag tag);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(const char* str);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(std::string str);
+
+  /** Dynamic cast to a response. */
+  const Response* toResponse() const;
+
+  /** Dynamic cast to a response, mutable. */
+  Response* toResponse();
+
+  /** Dynamic cast to other message subclasses, const and non-const. */
+  #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+  const class name__* to##name__() const; \
+  class name__* to##name__();
+  #include "./message-macro-list.h"
+  #undef CHPLDEF_MESSAGE
+
+  /** Base visitor over messages, does nothing and returns default value. */
+  template <typename T>
+  class Visitor {
+  private:
+    Server* ctx_;
+
+  public:
+    Visitor(Server* ctx) : ctx_(ctx) {} 
+
+    virtual T visit(const class Response* rsp) {
+      T ret;
+      return ret;
+    }
+
+    /** Default visitor body is trivial enough that we define it here. */
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+    virtual T visit(const class name__* req) { \
+      T ret; \
+      return ret; \
+    }
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+
+    /** Invoke the visitor on a given message, defined below. */
+    T dispatch(const Message* req);
+
+    /** Get the server context. */
+    Server* ctx() const;
+  };
+
+  /** Dispatch a visitor using a message as the receiver. */
+  template <typename T, typename U>
+  inline T accept(Message::Visitor<U>& v) { return v.dispatch(*this); }
+
+  /** Some messages can defer, but not sure which at this moment. */
+  virtual bool defer() const { return false; }
+
+  /** Pack this message into a JSON value. */
+  virtual JsonValue pack() const = 0;
+
+  /** May handle a message, but does not send or receive. */
+  static opt<Response> handle(Server* ctx, Message* msg);
+};
+
+/** Most messages are incoming and are modelled as a 'Request'. A request
+    consists of a set of parameter values, that can be used to compute
+    a result (it is analogous to a function call). Some requests can be
+    sent by the server to the client, in which case they are 'sendable',
+    and should provide a definition for the 'pack()' method.
+*/
+template <typename Params, typename Result>
+class Request : public Message {
+public:
+  /** Stores computation results for this request. */
+  struct ComputedResult {
+    bool isProgressingCallAgain;
+    Message::Error error;
+    std::string note;
+    Result result;
+  };
+
+protected:
+  const Params p;
+  Result r = {};
+
+  static Message::Error unpack(const JsonValue& json, Params& p,
+                               std::string* note);
+
+  Request(Message::Tag tag, JsonValue&& id, Message::Error error,
+          std::string&& note,
+          Params&& params)
```

Comment:
I don't think you need to accept by rvalue reference here. The common pattern is accept by value and move.

Reply:
I've made this change across the codebase.